sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/client/SimplifiedKinesisClient.java

```diff
@@ -0,0 +1,164 @@
+
+                return shards;
+            }
+        });
+    }
+
+    /***
+     * Gets records from Kinesis and deaggregates them if needed.
+     *
+     * @return list of deaggregated records
+     * @throws IOException - in case of recoverable situation
+     */
+    public GetKinesisRecordsResult getRecords(String shardIterator, String streamName,
+                                              String shardId) throws TransientKinesisException {
+        return getRecords(shardIterator, streamName, shardId, null);
+    }
+
+    /***
+     * Gets records from Kinesis and deaggregates them if needed.
+     *
+     * @return list of deaggregated records
+     * @throws IOException - in case of recoverable situation
+     */
+    public GetKinesisRecordsResult getRecords(final String shardIterator, final String streamName,
+                                              final String shardId, final Integer limit)
+            throws
+            TransientKinesisException {
+        return wrapExceptions(new Callable<GetKinesisRecordsResult>() {
+            @Override
+            public GetKinesisRecordsResult call() throws Exception {
+                GetRecordsResult response = kinesis.getRecords(new GetRecordsRequest()
+                        .withShardIterator(shardIterator)
+                        .withLimit(limit));
+                return new GetKinesisRecordsResult(
+                        UserRecord.deaggregate(response.getRecords()),
+                        response.getNextShardIterator(),
+                        streamName, shardId);
+            }
+        });
+    }
+
+    /***
+     * Wraps Amazon specific exceptions into more friendly format.
+     *
+     * @throws IOException              - in case of recoverable situation, i.e.
+     *                                  the request rate is too high, Kinesis remote service
+     *                                  failed, network issue, etc.
+     * @throws ExpiredIteratorException - if iterator needs to be refreshed
+     * @throws RuntimeException         - in all other cases
+     */
+    private <T> T wrapExceptions(Callable<T> callable) throws TransientKinesisException {
+        try {
+            return callable.call();
+        } catch (ExpiredIteratorException e) {
+            throw e;
+        } catch (LimitExceededException | ProvisionedThroughputExceededException e) {
+            LOG.warn("Too many requests to Kinesis", e);
+            throw new TransientKinesisException(e);
+        } catch (AmazonServiceException e) {
+            if (e.getErrorType() == AmazonServiceException.ErrorType.Service) {
+                LOG.warn("Kinesis backend failed", e);
+                throw new TransientKinesisException(e);
+            }
+            LOG.error("Client side failure", e);
+            throw new RuntimeException(e);
+        } catch (Exception e) {
+            LOG.error("Unknown failure", e);
+            throw new RuntimeException(e);
```

Comment:
Can you include the log messages in the created exceptions?

(Also, maybe, maybe the caller could choose whether to log.)